File: katas/csharp/Patterns/factory/factory-good.cs
using System;

// Implementación del Patrón Factory: Creación centralizada de objetos
// ✅ Solución: Crear una fábrica que maneje la creación de objetos

// Interfaz común ✅
public interface INotification
{
  string Send(string message);
}

// Implementaciones concretas ✅
public class EmailNotification : INotification
{
  public string Send(string message)
  {
    return $"EMAIL: {message}";
  }
}

public class SMSNotification : INotification
{
  public string Send(string message)
  {
    return $"SMS: {message}";
  }
}

public class PushNotification : INotification
{
  public string Send(string message)
  {
    return $"PUSH: {message}";
  }
}

// ✅ La fábrica centraliza la creación de objetos
public class NotificationFactory
{
  public static INotification Create(string type)
  {
    switch (type)
    {
      case "email":
        return new EmailNotification();
      case "sms":
        return new SMSNotification();
      case "push":
        return new PushNotification();
      default:
        throw new Exception($"Tipo de notificación desconocido: {type}");
    }
  }
}

// ✅ El código cliente no conoce las clases concretas
public class NotificationService
{
  public string SendNotification(string type, string message)
  {
    // ✅ La fábrica maneja la creación de objetos
    var notification = NotificationFactory.Create(type);
    return notification.Send(message);
  }
}

// ✅ Beneficios:
// 1. Creación centralizada de objetos
// 2. Fácil agregar nuevos tipos de notificación (solo actualizar la fábrica)
// 3. El código cliente depende de la interfaz, no de clases concretas

class Program
{
  static void Main()
  {
    Console.WriteLine("=== Solución con Patrón Factory ===");
    var service = new NotificationService();
    Console.WriteLine(service.SendNotification("email", "Hola Mundo"));
    Console.WriteLine(service.SendNotification("sms", "Hola Mundo"));
    Console.WriteLine(service.SendNotification("push", "Hola Mundo"));

    // ¡Agregar un nuevo tipo es fácil - solo cambiar la fábrica!
    Console.WriteLine("\n=== Nuevo tipo agregado ===");
    // Para agregar SlackNotification, solo crear la clase e incluir en la factory
  }
}

// ✅ Agregar un nuevo tipo (ejemplo):
public class SlackNotification : INotification
{
  public string Send(string message)
  {
    return $"SLACK: {message}";
  }
}

File: katas/java/Patterns/factory/FactoryGood.java
package Patterns.factory;

// Implementación del Patrón Factory: Creación centralizada de objetos
// ✅ Solución: Crear una fábrica que maneje la creación de objetos

public class FactoryGood {
    public static void main(String[] args) {
        System.out.println("=== Solución con Patrón Factory ===");
        NotificationService service = new NotificationService();
        System.out.println(service.sendNotification("email", "Hola Mundo"));
        System.out.println(service.sendNotification("sms", "Hola Mundo"));
        System.out.println(service.sendNotification("push", "Hola Mundo"));
    }
}

// Interfaz común ✅
interface Notification {
    String send(String message);
}

// Implementaciones concretas ✅
class EmailNotification implements Notification {
    public String send(String message) {
        return "EMAIL: " + message;
    }
}

class SMSNotification implements Notification {
    public String send(String message) {
        return "SMS: " + message;
    }
}

class PushNotification implements Notification {
    public String send(String message) {
        return "PUSH: " + message;
    }
}

// ✅ La fábrica centraliza la creación de objetos
class NotificationFactory {
    public static Notification create(String type) {
        switch (type) {
            case "email":
                return new EmailNotification();
            case "sms":
                return new SMSNotification();
            case "push":
                return new PushNotification();
            default:
                throw new IllegalArgumentException("Tipo de notificación desconocido: " + type);
        }
    }
}

// ✅ El código cliente no conoce las clases concretas
class NotificationService {
    public String sendNotification(String type, String message) {
        // ✅ La fábrica maneja la creación de objetos
        Notification notification = NotificationFactory.create(type);
        return notification.send(message);
    }
}

// ✅ Beneficios:
// 1. Creación centralizada de objetos
// 2. Fácil agregar nuevos tipos de notificación (solo actualizar la fábrica)
// 3. El código cliente depende de la interfaz, no de clases concretas
